### ISSUE TITLE
devtools triggering action multiple times

### ISSUE THREAD

**Issue by coodoo** [gaearon/redux-devtools#32]
Seems devtools is triggering every `action` multiple times, have a look at the sample code [here](https://github.com/coodoo/react-redux-isomorphic-example/tree/devtools-issue).

After starting the app, click "Add to cart" on "H&M T-Shirt White - 10.99" twice, and notice the 2nd time there are 2 items instead of 1 added to the cart, then click the button again, this time 3 items will be added.

I've traced it down to that devtools is triggering the action (`ADD_TO_CART`) multiple times, once devtools is disabled the issue is gone.

I'm wondering is this a bug in devtool or am I using it wrong? Thanks.

**Comment by FredyC** [comment]
I am not sure if it's the same thing, but devtools are always replaying all the actions from the beginning when new one is added. That's the feature, not a bug :)

**Comment by tomsdev** [comment]
I spent a lot of time today investigating the same problem (actions being called twice) when writing a custom middleware with redux-devtools enabled. Maybe you should make that behavior more explicit via a parameter that could also be set to false? I was expecting actions to be replayed only when the code is updated.

**Comment by gaearon** [comment]
@tomsdev

If you have problems because of it, this means your reducers aren't pure, and it's a problem in your app.

This means you are not using Redux correctly and you'll have other problems too, like views not updating in some cases.

See my comments and detailed description of the problem here: https://github.com/coodoo/react-redux-isomorphic-example/issues/9. (Author of this issue has already fixed that problem in their code.)

**Comment by cpsubrian** [comment]
The devtools are working as expected (replaying all actions), but this had some unfortunate consequences for me due to heavy usage of `<CSSTransitionGroup>` for my "super fancy ui" :smile:

I'll need to dive into the devtools more to understand why replaying all the actions every time is a "feature", but for my purposes its causing all sorts of *visual* havok. The app always ends in the correct state though, its just hell to look at.

**Comment by gaearon** [comment]
The first step towards fixing this is writing tests. :-) Recomputing every time is not a feature. I'm just not happy trying to optimize things before we get some tests. If you feel like contributing some, let us know!

**Comment by cpsubrian** [comment]
Oh for sure will contribute once I wrap my head around it. Just got it all working an hour ago :smile:

**Comment by gaearon** [comment]
Wait, I'm confused now. Redux DevTools replays actions against a reducer but it will dispatch update just once. This shouldn't affect your UI code at all.

Can you put up a project reproducing this problem?

**Comment by cpsubrian** [comment]
Alright, dug into the debugger and traced through whats happening. You're right what I thought was UI 'thrashing' was indeed just one update (there are a lot of animations). The problems lies in the fact that my reducers don't return the exact same state given the same inputs, but in a subtle way I'm not sure exactly how to solve.

The UI builds up a grid of rows and cells. There are actions to add an empty row, and an empty cell to a row. Currently, when you add a new row or cell they are assigned a randomly-generated ID. This is then used as the `key` such that React can properly track them when they move or are inserted or deleted.

Since the Devtools rerun all the reducers, you end up with the same grid, but all the ids are different. This causes React to re-render everything and trigger all the css transitions.

In a nutshell, I need to figure out a way to generate unique IDs in a way that when the reducers are re-run you end up with the same ids. Basically the ids need to depend on the state with some sort of auto-increment or other predictable pattern.

**Comment by gaearon** [comment]
Can you generate IDs in action creators somehow? Or are there just too many IDs to generate? Can you make this deterministic with some kind of seed value?

**Comment by cpsubrian** [comment]
Moving the id generation to my action creators fixed the issue, though I'll admit it doesn't feel as clean. I'm used to actions being more or less identity functions, but probably this is something I just need to get used to. In other flux implementations I've used there is sometimes quite a bit of logic in the 'stores', but I can see that with redux you want keep reducers as only 'calculators'  (as the docs put it). Just a tiny paradigm shift to accept.

**Comment by gaearon** [comment]
Yeah. It's a bit of a different paradigm. It's important to keep reducers pure than anything else. Record/replay is impossible with impure reducers.

**Comment by FredyC** [comment]
@cpsubrian I also had ID generation in a reducers, it seemed logical back there. Than I've moved it to action creators because I had same issues as you. Now I've come to situation where I need to know what ID was generated so I can dispatch another action with it. Since dispatch cannot provide such feedback, I have to move ID generation before actual dispatch. Basically I am creating another layer that can prepare data for action creators. 

I want to do that for most of the actions eventually. It 's starting to make sense that reducers should be about transforming action data into state change without any conditions and hidden logic. That way it would be more clear from just action log to see what actually happened there.

**Comment by marcopeg** [comment]
Hi guys, I believe this thread is an important part of the redux doc as it describes a common mistakes we can fall into when we start "thinking redux".

I came to this page 'cause I've used time based generated ids in a reducer. As @FredyC said, it seems quite natural to me. But in the end that simple mistake makes the reducer being no pure anymore.

I suggest to dedicate a page to ids generation within action creators, explaining what is the conceptual error and serving some good examples.

As for me, moving the id: `Date.now()` from the reducer to the action creator solved the problem completely. It may appear silly but it drove me crazy yesterday :-)

Thank you for this page!

**Comment by vincent-dm** [comment]
I was also bitten by this one. I believe that my reducer is pure, since it always returns the same object if the previous state and action are identical:
```js
return Object.assign(
        {},
        state,
        {
          fileTypes: action.fileTypes.map(calculateNestedFilecount),
        }
      )
```

The action passes a nested tree-like structure from the server. It contains a value for each "leaf", but the values for the branches are calculated by `calculateNestedFilecount` (I see no other logical place to do this).

Thus, I modify the action object being passed to my reducer, but not the state. I was under the impression that this was allowed in redux, since actions are supposed to be ephemeral.

Shouldn't redux-devtools use a clone of the action when replaying it?

**Comment by gaearon** [comment]
We can't do that because we don't know how deep to clone. With a shallow clone you'll have the same problem. Also, deep cloning is a lossy operation if object identities are important. And quite expensive too.

A function is not pure if it modifies either its argument. Modifying action indeed makes a reducer impure. Don't do this ;-)

**Comment by coodoo** [comment]
Well said, we learned that the hard way 😎

**Comment by vincent-dm** [comment]
Ok, I see. Thanks for the explanation @gaearon. I erroneously thought that Object.assign() copied the entire object.

**Comment by rpedroni** [comment]
Hey guys, I'm building a simple SimonSays with Redux and React and have a need for random number generation. I've currently put the random number generation in the action creator, just like some of you have suggested (and seemed quite natural to do, actually). Question is: the docs (http://rackt.github.io/redux/docs/basics/Actions.html) clearly state that creators should be pure, too. Any thoughts on this? And if so, what should be a better place to do this?

**Comment by marcopeg** [comment]
@rpedroni my personal take is that **reducers must be pure** but action creators don't strictly need too. But if they are pure everything is way more simple to test.

My approach is to use the `redux-thunk` middleware to implement **asynchronous actions** which I call "services". Those actions are in charge of everything which is not a pure function:

- async stuff (ajax, timing)
- unpredictable results (aka randomisation)
- failure prone actions (like localStorage read/write)
- pub/sub

Here is an example:

```
// import the pure action creator to set a value into the store
var setDiceValue = require('pure-actions/set-dice-value');

// define a "service" to generate the random number
function myRandomAction() {
  return function(dispatch, getState) {
    dispatch(setDiceValue(random(1, 6)));
  }
}
```

### Advantages:

- you keep as many pure actionCreators as you can so it will be a breeze to write quick regression tests for your in-app apis
- you store into "services" all the impure code and this code will interface with the rest of your app dispatching pure actions.

In the long run "services" will host all your business logic. The fact that they receive a simple "dispatch, getState" interface to interact with the rest of the application makes testing services very well scoped.

Please let me know how it goes. 
Marco

**Comment by rpedroni** [comment]
Thanks for the detailed answer, @marcopeg ! For now, I'm going with the impure action creators since this is more of a demo project than anything else and I'm unit testing the this creator, despite it's randomness, by expecting an integer with range between 0 and N (hell, good enough for this project :))

But I'll definitely try it your way on my next larger project. Thanks!
